fix: Move Hunger Games to a real view instead of overlay (#4681)

LorenzoMascia opened the pull request:
Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes :
	modified:   packages/smooth_app/lib/pages/hunger_games/question_card.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

### What
- Moved the Hunger Games screen from an overlay to a real view using a standard `Scaffold`, as described in the issue.
- This improves navigation consistency and aligns the UI with the rest of the app.

**fix: Move Hunger Games to real view instead of overlay**

### Screenshot

![Screenshot 2025-04-15 211841](https://github.com/user-attachments/assets/49474f14-a649-4a50-acbe-81154698bd93)
![Screenshot 2025-04-15 212004](https://github.com/user-attachments/assets/d0eb2634-9948-4920-9a0e-13fadaeb880c)

### Fixes bug(s)
- Fixes: #4681

Commit:
Fix #4681: Move Hunger Games to a real view instead of overlay

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes :
	modified:   packages/smooth_app/lib/pages/hunger_games/question_card.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

monsieurtanuki commented on packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart:
```diff
@@ -308,7 +308,17 @@ class UserPreferencesContribute extends AbstractUserPreferences {
+        builder: (context) => QuestionsPage( ),
+        ),
+      ).then((questionsAnswered) {
+      // Gestisci il risultato se necessario
```
Per favore, niente in italiano.

LorenzoMascia replied:
done, no more Italian

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_questions_widget.dart:
```diff
@@ -77,20 +77,24 @@ class _ProductQuestionsWidgetState extends State<ProductQuestionsWidget>
     _trackEvent(AnalyticsEvent.questionClicked);
 
-    final int? answeredQuestions = await openQuestionPage(
+    Navigator.push(
```
No `await`? Then please make it explicit and use `unawait(`.

LorenzoMascia replied:
done

monsieurtanuki commented:
Hi @LorenzoMascia!
First thing, have your code formatted (cf. `flutter format .` and CTRL-S).

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -445,12 +363,13 @@ class _QuestionsSuccessView extends StatelessWidget {
-              _QuestionPageState._numberQuestionsNext,
+          ? AppLocalizations.of(context).robotoff_next_n_questions(10
+              
+              // QuestionPageState._numberQuestionsNext,
```
No useless comments, please.

LorenzoMascia replied:
removed the comment and replaced the 10 with the variable numberQuestionsNext

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -367,7 +281,7 @@ class _LoadingQuestionsView extends StatelessWidget {
             child: DefaultTextStyle(
               textAlign: TextAlign.center,
-              style: const TextStyle(),
+              style: TextStyle(color: colorScheme.onBackground), 
```
Not sure what makes you use explicit colors.
Anyway, if you select `onBackground` as a background color, shouldn't you use `background` as a front color?
Shouldn't you rather use `primary` and `onPrimary`?
Or no color at all, just using the app and the device's default values?

LorenzoMascia replied:
used primary, and I had to specify the color because with the light theme the lettering was not visible

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +96,94 @@ class _QuestionPageState extends State<_QuestionPage>
+
+  void _incrementQuestionsAnswered() {
+    setState(() {
+      _questionsAnswered++;
```
Not sure how `setState` works but `++_questionsAnswered` would be more appropriate.

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +96,94 @@ class _QuestionPageState extends State<_QuestionPage>
+      create: (context) => _QuestionsAnsweredNotifier(),
+      child: Scaffold(
+        appBar: AppBar(
+          title: Text("Hunger Games"), //AppLocalizations.of(context).hunger_games_error_label),
```
No useless comments please.

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_card.dart:
```diff
@@ -62,7 +62,7 @@ class QuestionCard extends StatelessWidget {
                 children: <Widget>[
                   SizedBox(
-                    height: screenSize.height / 6,
+                    height: screenSize.height / 4,
```
Would an `Expanded` give us more flexibility?

LorenzoMascia replied:
using `Expanded `
![Screenshot 2025-04-16 182409](https://github.com/user-attachments/assets/854fa1c7-0bf7-472b-8456-bffa584d5a9d)
![Screenshot 2025-04-16 182611](https://github.com/user-attachments/assets/53b5ba78-c3a0-4cd1-afcb-4d16972412d2)

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +96,94 @@ class _QuestionPageState extends State<_QuestionPage>
+      );
+      _currentQuestionIndex = 0;
+    } catch (e) {
+      debugPrint('Error loading next questions: $e');
```
No `debugPrint`, please.

LorenzoMascia replied:
`debugPrint` removed

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +96,94 @@ class _QuestionPageState extends State<_QuestionPage>
-    );
-    _currentQuestionIndex = 0;
+    try {
+      setState(() {
```
Generally speaking, you're supposed to test `if(mounted)` before calling `setState`.

monsieurtanuki replied:
ping

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +96,94 @@ class _QuestionPageState extends State<_QuestionPage>
+    }
+  }
+
+  void _incrementQuestionsAnswered() {
```
Cannot see the added value of moving the code here.
It was more relevant to put that in `_saveAnswer`, wasn't it?

LorenzoMascia replied:
function removed

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_questions_widget.dart:
```diff
@@ -77,20 +77,24 @@ class _ProductQuestionsWidgetState extends State<ProductQuestionsWidget>
+            .toList(growable: false),
+            ),
+          ),
+        ).then((questionsAnswered) {
```
We don't use that much the `then` syntax. Harder to read.
Please stick to the `int? value = await myMethod(); if (value ...` syntax.

LorenzoMascia replied:
changed the construct

monsieurtanuki reviewed (commented):
Hi @LorenzoMascia!
Of course the first thing is to fix the formatting, but please read my other comments.

Commit:
Fixes after code review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   packages/smooth_app/lib/pages/hunger_games/question_card.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_questions_widget.dart:
```diff
@@ -77,16 +77,20 @@ class _ProductQuestionsWidgetState extends State<ProductQuestionsWidget>
     _trackEvent(AnalyticsEvent.questionClicked);
 
-    final int? answeredQuestions = await openQuestionPage(
+    final int? questionsAnswered = await Navigator.push<int>(
```
For the record, don't change the naming of the variables unless the change is very relevant.
It was probably not the case here, and the reviewer still has to check every change...

LorenzoMascia replied:
Changed

monsieurtanuki commented on packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart:
```diff
@@ -308,7 +308,13 @@ class UserPreferencesContribute extends AbstractUserPreferences {
     );
-    await openQuestionPage(context);
+
+    Navigator.push(
```
Please use the simpler `await` syntax.

LorenzoMascia replied:
Done

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -416,18 +327,22 @@ class _QuestionView extends StatelessWidget {
-      ],
+    return Expanded(
+      child: Column(
+        mainAxisSize: MainAxisSize.max, // <--- CAMBIATO
```
Per favore...

LorenzoMascia replied:
deleted, sorry

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart:
```diff
@@ -10,31 +10,43 @@ class QuestionImageThumbnail extends StatelessWidget {
+                fit: BoxFit.cover,
+                errorBuilder: (_, __, ___) => EMPTY_WIDGET,
+                loadingBuilder: (
+                  BuildContext context,
+                  Widget child,
+                  ImageChunkEvent? loadingProgress,
+                ) {
+                  if (loadingProgress == null) {
+                    // TODO(monsieurtanuki): remove this when the image is not null anymore
+                    return child;
+                  }
+                  return const Center(
+                      child: CircularProgressIndicator.adaptive());
+                },
+              ),
+            );
+          },
```
Honestly, what's the point of changing the parameter names and the coding style, except annoying the reviewer?
You literally changed nothing here, except adding an unexpected `TODO` with my name...
Please revert that part.

LorenzoMascia replied:
Reverted

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart:
```diff
@@ -10,31 +10,43 @@ class QuestionImageThumbnail extends StatelessWidget {
-                progress == null
-                    ? child
-                    : const CircularProgressIndicator.adaptive(),
+  Widget build(BuildContext context) {
+    return Container(
```
What's the point of changing the coding style? Please revert that part.

LorenzoMascia replied:
Reverted

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart:
```diff
@@ -10,31 +10,43 @@ class QuestionImageThumbnail extends StatelessWidget {
+        child: LayoutBuilder(
+          builder: (BuildContext context, BoxConstraints constraints) {
+            return SizedBox(
+              width: constraints.maxWidth,
+              height: constraints.maxHeight,
```
Would you please try with `double.infinity` for width and height?

LorenzoMascia replied:
Changed

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -18,119 +16,9 @@ import 'package:smooth_app/query/product_questions_query.dart';
-  const _QuestionPage({
+class QuestionsPage extends StatefulWidget {
+  const QuestionsPage({
+    super.key,
```
We don't need `key`, do we?

LorenzoMascia replied:
No, we don't. Deleted.

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -205,70 +94,90 @@ class _QuestionPageState extends State<_QuestionPage>
+      setState(() {
+        _state = const _RobotoffQuestionLoadingState();
+      });
+      _loadQuestions(
```
Please add `unawaited(`.

LorenzoMascia replied:
Added

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -367,7 +274,7 @@ class _LoadingQuestionsView extends StatelessWidget {
             child: DefaultTextStyle(
               textAlign: TextAlign.center,
-              style: const TextStyle(),
+              style: TextStyle(color: colorScheme.onSurface),
```
Please
* make a screenshot
* remove all your color and style changes, and then make a screenshot

LorenzoMascia replied:
You are right, the screenshots are exactly the same.

monsieurtanuki reviewed (commented):
Hi @LorenzoMascia, and thank you for your changes!
I like the `Expanded` version, though the image size isn't correctly computed (the bottom of the big image isn't visible). That needs to be fixed (maybe removing the `SizedBox` and letting the system compute by itself, I don't know).
Please read my other comments.

Commit:
Fixes after second review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   packages/smooth_app/lib/pages/hunger_games/question_page.dart
	modified:   packages/smooth_app/lib/pages/preferences/user_preferences_contribute.dart
	modified:   packages/smooth_app/lib/pages/product/product_questions_widget.dart

LorenzoMascia commented:
The final result using BoxFit.contain instead of cover for the ìmage:

![Screenshot 2025-04-21 202721](https://github.com/user-attachments/assets/c7e248fe-c218-4f73-a644-d24313ff034e)

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_image_thumbnail.dart:
```diff
@@ -21,19 +21,26 @@ class QuestionImageThumbnail extends StatelessWidget {
-                progress == null
-                    ? child
-                    : const CircularProgressIndicator.adaptive(),
+          child: LayoutBuilder(
```
Then we don't need a `LayoutBuilder` anymore, do we?

LorenzoMascia replied:
No, it's no longer needed. I had used the builder to get the BoxConstraints during some trials, but in the end I didn’t use them.

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_card.dart:
```diff
@@ -59,10 +57,10 @@ class QuestionCard extends StatelessWidget {
-                  SizedBox(
-                    height: screenSize.height / 6,
+                  Expanded(
+                    flex: 1,
```
```suggestion
```
Not needed, is it?

monsieurtanuki commented on packages/smooth_app/lib/pages/hunger_games/question_page.dart:
```diff
@@ -416,18 +325,22 @@ class _QuestionView extends StatelessWidget {
-          onAnswer: onAnswer,
-        ),
-      ],
+    return Expanded(
```
Not sure if that `Expanded` is relevant.

LorenzoMascia replied:
No, it works in the same way returning directly the Column without the Expanded

monsieurtanuki reviewed (approved):
Good job @LorenzoMascia!

If you have 10 minutes, please have a look at my comments. If you already know that you won't have time for that in a near future, please tell me.

I'm still surprised that you had to change the styles and colors, but that's another story.

monsieurtanuki commented:
@LorenzoMascia And congratulations for your first PR here!

LorenzoMascia committed:
Update packages/smooth_app/lib/pages/hunger_games/question_card.dart

Co-authored-by: monsieurtanuki <fabrice_fontaine@hotmail.com>

LorenzoMascia committed:
Merge branch 'develop' into fix/hunger-games-view

Commit:
Fixes after third review

Committer: Lorenzo Mascia <lorenzo.mascia@ericsson.com>

Changes:
	modified:   lib/pages/hunger_games/question_image_thumbnail.dart
	modified:   lib/pages/hunger_games/question_page.dart

monsieurtanuki commented:
Hi @LorenzoMascia!
Before I merge your code, would you please attach a screenshot?
I'm a bit worried about your choice of adding extra styles and extra colors. Thank you!

LorenzoMascia commented:
### **Screenshot with App Setting->Theme Light:**
![Screenshot 2025-04-27 175756](https://github.com/user-attachments/assets/7db877da-602a-4516-acf1-495aede6136b)

### **Screenshot with App Setting->Theme Dark:**
![Screenshot 2025-04-27 175656](https://github.com/user-attachments/assets/3609fd58-7df0-464a-968c-1523caa2fa58)

monsieurtanuki commented:
Thank you @LorenzoMascia!

Now, would you please attach the very same screenshots _without_ your text size, style and color modifications? I'm really puzzled about how relevant they are. Especially as in both dark and light modes:
* we use the same foregrounds and backgrounds for all "hunger games specific" items (e.g. the "YES" button), therefore we don't need any tap dancing about `onPrimary` and `primary`, it's just always white on green.
* we use the default colors for the rest (e.g. the name of the product)

Besides, I'm a bit nervous about the instability of code like this one:
```dart
            style: textTheme.bodyLarge?.copyWith(
              fontWeight: FontWeight.bold,
              fontSize: 20.0,
              color: colorScheme.onSurface,
            ) ??
            const TextStyle(),
```
So, if there is no `textTheme.bodyLarge`, we forget about setting `fontWeight`, `fontSize` and `color` altogether.

Sorry about that.

LorenzoMascia commented:
**Without restyling there is an issue with the light theme. Following the screenshots of Loading and Error page**

![Loading Default Styles Light](https://github.com/user-attachments/assets/dfe6e474-72cf-4597-8c71-5691a3b5c169)
![Error Light Defaukt Style](https://github.com/user-attachments/assets/84d7c78b-cd3d-4492-b5a0-562cee565937)

**The same screens are ok in the dark theme** 

![Loading Defaukt Styles Dark](https://github.com/user-attachments/assets/832f4cff-5876-40d9-8686-46878e75d6fb)
![Error Dark Defaul Style](https://github.com/user-attachments/assets/0a367634-e074-4c70-8f7b-ab182ac2685c)


**Following the screens with style override.**

![Screenshot 2025-04-27 175736](https://github.com/user-attachments/assets/eb2e766d-a547-4333-b056-a96064455cf0)
![Screenshot 2025-04-28 210404](https://github.com/user-attachments/assets/5780f76a-859b-4e3a-a31b-85f2c0d32129)
![Loading restyled Dark](https://github.com/user-attachments/assets/9cc22af6-55a4-4d9b-a333-8c8644c61ed9)
![Screenshot 2025-04-28 210225](https://github.com/user-attachments/assets/96a1625c-326e-4be9-8815-91d6b07dbf4c)

No styles applied to the other pages

monsieurtanuki commented:
Thank you @LorenzoMascia!
